### ISSUE TITLE
[feature] Create lock files based on date and time

### ISSUE THREAD

**Issue by lpapp-foundry** [conan-io/conan#14539]
### What is your suggestion?

conan lock create has a --update parameter

What would be good is that based on a past commit date and time in git, we could generate lock files for that commit, the nearest revision of conan dependency recipes earlier than the git commit.

This would be e.g. very helpful in projects where these used to be no lock files at a certain point of time, so builds would be otherwise difficult to reproduce.

At the moment, I end up checking all the dependencies manually to pin them by conan -r foo -rev Bar/X.YZ@... and then manually pick the corresponding pin from the list, but it is struggle to do this manually for all the dependencies. It is also error prone.

It sounds like this could be automated?

### Have you read the CONTRIBUTING guide?

- [ ] I've read the CONTRIBUTING guide

**Comment by memsharded** [comment]
Hi @lpapp-foundry 

Thanks for your comments.

I am afraid that I don't understand the request.
When Conan resolves packages, it resolves them against a Conan remote like ``conan_server`` or ``Artifactory``. It fetches the recipes, evaluate them, construct the graph, and if generating a lockfile, saves the versions and revisions in the file. A lockfile is nothing but a snapshot of the current dependency graph versions and revisions.
Git repos are not involved at all in this process, no git repos are cloned, no commits are referenced. Everything is downloaded from the Conan remote. So it is not possible to do something like the "earlier commit", because there is no commit at all to start with. 
Could you please clarify your request?

**Comment by lpapp-foundry** [comment]
> Hi @lpapp-foundry
> 
> Thanks for your comments.
> 
> I am afraid that I don't understand the request. When Conan resolves packages, it resolves them against a Conan remote like `conan_server` or `Artifactory`. It fetches the recipes, evaluate them, construct the graph, and if generating a lockfile, saves the versions and revisions in the file. A lockfile is nothing but a snapshot of the current dependency graph versions and revisions. Git repos are not involved at all in this process, no git repos are cloned, no commits are referenced. Everything is downloaded from the Conan remote. So it is not possible to do something like the "earlier commit", because there is no commit at all to start with. Could you please clarify your request?

I was trying to propose a --date-time argument to the conan lock file generation and I will supply that argument from the `git show -s --format=%ci HEAD` output of my software using conan to fetch its dependencies. Conan itself does not have to know about git. But it would need to accept a date and time variable for generating the lock file rather than always the latest. Since that will not necessarily work for old revisions of my software.

So, not to generate lock files for the latest recipe versions, but the recipe versions nearest earlier that the date and time I am specifying based on the timestamp of the git commit of my software. Git is for my software, not for conan.

As written, I have been doing this process manually, but it seems to be time-consuming and error-prone.

If conan could automate it, it would be great.

conan lock create --update --date-time "02/12/2022" (or whatever time format your prefer) would generate lock files for all dependencies in the conanfile.py before this date, and the nearest recipe version.

It is "just" automating the manual look-up of each dependency as written:

conan -rev -r foo Bar/X.Y.Z@.../...
-> abcdef 03/12/2022
-> cbadev 01/12/2022 -> pick this for the lock file.
-> defbab  11/12/2021

And so on for each dependency. Many projects did not use to have lock files from the beginning of the project. This makes regression testing, e.g. with git bisect extremely challenging on an old revision due to lack of lock files for the time. But if we could automate the recreation of lock files based on an old commit date we are at, that would be helpful. Since those old past commits will not necessarily work with the latest recipe versions. More likely that they will not than they will.

So, to summarise, to relock an old version, which did not use lock files for them, so that I can rebuild it with exactly the same conan dependency package versions that were used for the time without lock files, so they were the latest at the time of that commit of my software.

Does this help or shall I provide even more context?

**Comment by memsharded** [comment]
Thanks for the clarification, now I understand the request better.
It has nothing to do with lockfiles, but this would be a graph-resolution thing, basically a parameter to ``conan install`` or ``conan graph info`` too, because those commands can also generate a lockfile. It is a constraint over the resolution of dependencies, similar to what a version-range is doing for versions.

This use case was not expected for lockfile creation for sure. The idea is that a lockfile is a snapshot of the current state of the dependency graph, to be able to repeat later in the future. It not intended as a wayback time machine, but only as a move forward thing.

This kind of feature could be more challenging that it could look like on the surface. For example, what revision should be picked when there are ``version-ranges`` defined? It should iterate all possible versions, and then every possible revision of every version? 

It seems that you are using Conan 1.X command. This kind of new feature will no longer be considered for Conan 1.X releases, so if anything, they should be discussed for adding to Conan 2.0.

As Conan 2.0 features a new dependency resolution algorithm, and also implement new lockfiles, and will require to use the latest recipes revisions from ConanCenter anyway, I think the way forward is capturing lockfiles from the start when lockfiles are desired. This feature proposal sounds too much for a flow that is not provided by any other package manager for other languages, because it is just intended for some legacy gap, but that shouldn't happen anymore moving forward

**Comment by memsharded** [comment]
Closing this ticket as unplanned, too much complexity for the value, and Conan 2 better lockfiles should also improve this.
